Sources/MapboxNavigation/NavigationViewHelper.swift

```diff
@@ -0,0 +1,170 @@
+
+import MapboxDirections
+import MapboxCoreNavigation
+import CoreLocation
+import UIKit
+
+
+class NavigationViewHelper: NavigationComponent, ViewCycleObserver {
+    
+    // MARK: - Properties
+    
+    weak var navigationViewData: NavigationViewData!
+    var annotatesSpokenInstructions = false
+    
+    fileprivate var navigationMapView: NavigationMapView {
+        return navigationViewData.navigationView.navigationMapView
+    }
+    
+    fileprivate var router: Router! {
+        navigationViewData.navigationService.router
+    }
+    
+    fileprivate var routeLineTracksTraversal: Bool {
+        get {
+            navigationMapView.routeLineTracksTraversal 
+        }
+        set {
+            navigationMapView.routeLineTracksTraversal = newValue
+        }
+    }
+    
+    private var currentLegIndexMapped = 0
+    private var currentStepIndexMapped = 0
+    
+    // MARK: - Public methods
+    
+    func populateName(for waypoint: Waypoint, populated: @escaping (Waypoint) -> Void) {
+        guard waypoint.name == nil else { return populated(waypoint) }
+        let location = CLLocation(latitude: waypoint.coordinate.latitude, longitude: waypoint.coordinate.longitude)
+        CLGeocoder().reverseGeocodeLocation(location) { (places, error) in
```

Comment:
It somehow escaped my attention that we’re still using Core Location to geocode the waypoint name when no name has been set by the developer or filled in by the Directions API. I’m not entirely sure we actually hit this code, since for better or worse, the Directions API normally fills in the street name as a fallback waypoint name. Even if it doesn’t, it should be fine to leave the name unset; the only consequence is that EndOfRouteViewController’s not-too-prominent waypoint name label would wind up empty.

Reply:
And even if we keep this method around, it has no dependency on the surrounding class, so it would be better to put it in an extension of the Waypoint class.

Reply:
I like the idea of removing this one :)